File: cc/032.cc
// Project Euler Problem 032
//
// Find the sum of all numbers that can be written as pandigital
// products.

#include <cstdio>
#include <set>
#include <cmath>
#include "mymath.h"

using namespace std;

int main(int argc, const char* argv[]) {
  int sum = 0;
  set<int> s;
  for (int i=1; i<100; i++) {
    if (i/10 == i%10) continue;
    if (i%10 == 0) continue;
    int di = digits(i);
    double l10i = log10(i);
    for (int j=i+1; log10(j)+l10i <= 5; j++) {
      int dj = digits(j);
      if (di&dj) continue;
      int dij = digits(i*j);
      if ((di|dj) & dij) continue;
      if ((di|dj|dij) == 0x3fe) {
        if (s.insert(i*j).second) {
          sum += i*j;
        };
      }
    }
  }
  printf("%d\n", sum);
}

File: go/problems/032.go
// Project Euler Problem 032
//
// Find the sum of all numbers that can be written as pandigital
// products.

package problems

import "fmt"

func Match(numbers ...int) bool {
	mask := 1 // 0 is out!
	for _, number := range numbers {
		for ; number > 0; number /= 10 {
			mask |= 1 << uint(number%10)
		}
	}
	return (mask == 0x3ff)
}

func Problem032() string {
	found := make(map[int]bool)
	// Lengths 1, 4, 4
	for a := 2; a <= 9; a++ {
		for b := 1234; b <= 9876; b++ {
			if a*b <= 9876 && Match(a, b, a*b) {
				found[a*b] = true
			}
		}
	}
	// Lengths 2, 3, 4
	for a := 12; a <= 98; a++ {
		for b := 123; b <= 987; b++ {
			if a*b <= 9876 && Match(a, b, a*b) {
				found[a*b] = true
			}
		}
	}
	sum := 0
	for k, _ := range found {
		sum += k
	}
	return fmt.Sprintf("%d", sum)
}

func init() {
	Functions[32] = Problem032
}
